### PR TITLE
JAMES-2873 Add Hostname into Created and Started events

### PR DESCRIPTION
**Pull request by mbaechler**

### CODE REVIEW

File: server/task/src/main/scala/org/apache/james/task/TaskExecutionDetails.scala [hunk 1]
```diff
@@ -50,7 +52,9 @@ class TaskExecutionDetails(val taskId: TaskId,
 
   def getAdditionalInformation: Optional[TaskExecutionDetails.AdditionalInformation] = additionalInformation()
 
-  def getSubmitDate: Optional[ZonedDateTime] = submitDate
+  def getSubmitDate: Optional[ZonedDateTime] = submittedDate
```
Comment (remk) [inline server/task/src/main/scala/org/apache/james/task/TaskExecutionDetails.scala]: s/getSubmitDate/getSubmittedDate

File: server/task/src/main/scala/org/apache/james/task/TaskExecutionDetails.scala [hunk 2]
```diff
@@ -100,47 +107,51 @@ class TaskExecutionDetails(val taskId: TaskId,
...
+      .add("startedDate", startedDate)
+      .add("ranNode", ranNode)
+      .add("completedDate", completedDate)
+      .add("canceledDate", canceledDate)
+      .add("failedDate", failedDate)
       .toString
 
   private def start = new TaskExecutionDetails(taskId, `type`, additionalInformation, IN_PROGRESS,
-    submitDate = submitDate,
+    submittedDate = submittedDate,
```
Comment (remk) [inline server/task/src/main/scala/org/apache/james/task/TaskExecutionDetails.scala]: add ran node

**Comment by trantienduchn** [comment]
```
[6ea2d9994422607e971eeb0f9d229c45f1048d8a] [ERROR]   CassandraTaskExecutionDetailsProjectionDAOTest.setUp:55 » InvalidQuery The col...
[6ea2d9994422607e971eeb0f9d229c45f1048d8a] [ERROR]   CassandraTaskExecutionDetailsProjectionDAOTest.setUp:55 » InvalidQuery The col...
[6ea2d9994422607e971eeb0f9d229c45f1048d8a] [ERROR]   CassandraTaskExecutionDetailsProjectionDAOTest.setUp:55 » InvalidQuery The col...
[6ea2d9994422607e971eeb0f9d229c45f1048d8a] [ERROR]   CassandraTaskExecutionDetailsProjectionDAOTest.setUp:55 » InvalidQuery The col...
[6ea2d9994422607e971eeb0f9d229c45f1048d8a] [ERROR]   CassandraTaskExecutionDetailsProjectionDAOTest.setUp:55 » InvalidQuery The col...
[6ea2d9994422607e971eeb0f9d229c45f1048d8a] [ERROR]   CassandraTaskExecutionDetailsProjectionDAOTest.setUp:55 » InvalidQuery The col...
[6ea2d9994422607e971eeb0f9d229c45f1048d8a] [ERROR]   TaskExecutionDetailsProjectionContract.listDetailsShouldReturnLastUpdatedRecords » InvalidQuery
[6ea2d9994422607e971eeb0f9d229c45f1048d8a] [ERROR]   TaskExecutionDetailsProjectionContract.listShouldReturnAllRecords » InvalidQuery
[6ea2d9994422607e971eeb0f9d229c45f1048d8a] [ERROR]   TaskExecutionDetailsProjectionContract.listShouldReturnEmptyWhenNone » InvalidQuery
[6ea2d9994422607e971eeb0f9d229c45f1048d8a] [ERROR]   TaskExecutionDetailsProjectionContract.loadShouldBeAbleToRetrieveASavedRecord » InvalidQuery
[6ea2d9994422607e971eeb0f9d229c45f1048d8a] [ERROR]   TaskExecutionDetailsProjectionContract.loadShouldReturnEmptyWhenNone » InvalidQuery
[6ea2d9994422607e971eeb0f9d229c45f1048d8a] [ERROR]   TaskExecutionDetailsProjectionContract.updateShouldUpdateRecords » InvalidQuery
[6ea2d9994422607e971eeb0f9d229c45f1048d8a] [ERROR]   DistributedTaskManagerTest.givenOneEventStoreTwoEventTaskManagersShareTheSameEvents:80 » InvalidQuery
```

File: server/task/src/test/java/org/apache/james/task/eventsourcing/EventSourcingTaskManagerTest.java [hunk 3]
```diff
@@ -19,37 +19,44 @@
...
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
 import scala.Function1;
 
 @ExtendWith(CountDownLatchExtension.class)
 class EventSourcingTaskManagerTest implements TaskManagerContract {
 
+    private static final String HOSTNAME = Hostname.apply("foo");
```
Comment (chibenwa) [inline server/task/src/test/java/org/apache/james/task/eventsourcing/EventSourcingTaskManagerTest.java]: I can't get why the type here is a string and not `Hostname`...

File: server/task/src/main/scala/org/apache/james/task/TaskExecutionDetails.scala (hunk 1 above)
Comment (blackheaven) [reply server/task/src/main/scala/org/apache/james/task/TaskExecutionDetails.scala]: it's used for serialization/webadmin

File: server/task/src/test/java/org/apache/james/task/eventsourcing/EventSourcingTaskManagerTest.java (hunk 3 above)
Comment (remk) [reply server/task/src/test/java/org/apache/james/task/eventsourcing/EventSourcingTaskManagerTest.java]: Scala value class don't be play nice when used in java. It as been reworked in a fixup. 
The Hostname class is now a simple scala case class. And the `HOSTNAME` const is now rightly typed.

**Comment by chibenwa** [comment]
test this please (restart)

File: server/task/src/main/scala/org/apache/james/task/TaskExecutionDetails.scala [hunk 4]
```diff
@@ -23,21 +23,23 @@ import java.time.ZonedDateTime
...
+  def from(task: Task, id: TaskId, hostname: Hostname) = new TaskExecutionDetails(id, task.`type`, () => task.details, WAITING, submittedDate = Optional.of(ZonedDateTime.now), submittedNode = Optional.of(hostname))
 }
 
 class TaskExecutionDetails(val taskId: TaskId,
                            private val `type`: String,
                            private val additionalInformation: () => Optional[TaskExecutionDetails.AdditionalInformation],
                            private val status: TaskManager.Status,
-                           private val submitDate: Optional[ZonedDateTime] = Optional.empty(),
+                           private val submittedDate: Optional[ZonedDateTime] = Optional.empty(),
+                           private val submittedNode: Optional[Hostname] = Optional.empty(),
```
Comment (remk) [inline server/task/src/main/scala/org/apache/james/task/TaskExecutionDetails.scala]: Do the submitted node and date need to be optional ?

**Comment by remk** [comment]
```
org.apache.james.mailrepository.cassandra.CassandraMailRepositoryWithFakeImplementationsTest
[8a275944c24b646639f3ea49e9c7571feb70cfb1] [INFO] 
[8a275944c24b646639f3ea49e9c7571feb70cfb1] [INFO] Results:
[8a275944c24b646639f3ea49e9c7571feb70cfb1] [INFO] 
[8a275944c24b646639f3ea49e9c7571feb70cfb1] [ERROR] Errors: 
[8a275944c24b646639f3ea49e9c7571feb70cfb1] [ERROR]   CassandraMailRepositoryTest.storingAndRemovingMessagesConcurrentlyShouldLeadToConsistentResult » Execution
[8a275944c24b646639f3ea49e9c7571feb70cfb1] [ERROR]   CassandraMailRepositoryTest.storingAndRemovingMessagesConcurrentlyShouldLeadToConsistentResult » Execution
[8a275944c24b646639f3ea49e9c7571feb70cfb1] [INFO] 
[8a275944c24b646639f3ea49e9c7571feb70cfb1] [ERROR] Tests run: 168, Failures: 0, Errors: 2, Skipped: 1
```

**Comment by remk** [comment]
test this please

**Comment by chibenwa** [comment]
I don't see the hostname in cancelRequested events anymore as a separated commit. Is this normal?

**Comment by blackheaven** [comment]
> anymore

No :'(

Good catch

File: server/container/guice/guice-common/src/main/java/org/apache/james/modules/server/TaskManagerModule.java [hunk 5]
```diff
@@ -19,16 +19,36 @@
...
+
 import org.apache.james.task.MemoryTaskManager;
 import org.apache.james.task.TaskManager;
+import org.apache.james.task.eventsourcing.Hostname;
 
 import com.google.inject.AbstractModule;
 import com.google.inject.Scopes;
 
 public class TaskManagerModule extends AbstractModule {
+    private static class UnconfigurableHostname extends RuntimeException {
```
Comment (rouazana) [inline server/container/guice/guice-common/src/main/java/org/apache/james/modules/server/TaskManagerModule.java]: usually Exception are suffixed by Exception, no?

File: server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/DistributedTaskManagerModule.java [hunk 6]
```diff
@@ -34,6 +38,11 @@
 import com.google.inject.Scopes;
 
 public class DistributedTaskManagerModule extends AbstractModule {
+    private static class UnconfigurableHostname extends RuntimeException {
+        UnconfigurableHostname(String message, Exception originException) {
```
Comment (rouazana) [inline server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/DistributedTaskManagerModule.java]: idem? And why defining twice the same exception?

**Comment by remk** [comment]
```
[e18a59aa2f852ae8e91a85b99660862948c7d7d6] [ERROR]   JamesCapabilitiesServerTest.startShouldFailWhenNoUniqueIDCapability:145 
[e18a59aa2f852ae8e91a85b99660862948c7d7d6] Expecting:
[e18a59aa2f852ae8e91a85b99660862948c7d7d6]   <com.google.inject.CreationException: Unable to create injector, see the following errors:
[e18a59aa2f852ae8e91a85b99660862948c7d7d6] 
[e18a59aa2f852ae8e91a85b99660862948c7d7d6] 1) A binding to org.apache.james.task.eventsourcing.Hostname was already configured at org.apache.james.modules.server.HostnameModule.configure(HostnameModule.java:41) (via modules: com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> org.apache.james.modules.CommonServicesModule -> org.apache.james.modules.server.TaskManagerModule -> org.apache.james.modules.server.HostnameModule).
[e18a59aa2f852ae8e91a85b99660862948c7d7d6]   at org.apache.james.modules.server.HostnameModule.configure(HostnameModule.java:40) (via modules: com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> org.apache.james.modules.CommonServicesModule -> org.apache.james.modules.server.TaskManagerModule -> org.apache.james.modules.server.HostnameModule)
[e18a59aa2f852ae8e91a85b99660862948c7d7d6] 
```

**Comment by remk** [comment]
```
[7410efc08ea20c80f6ea22a5511f1055bc762ad9] [INFO] Tests run: 6, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.041 s - in org.apache.james.mailrepository.cassandra.CassandraMailRepositoryWithFakeImplementationsTest
[7410efc08ea20c80f6ea22a5511f1055bc762ad9] [INFO] 
[7410efc08ea20c80f6ea22a5511f1055bc762ad9] [INFO] Results:
[7410efc08ea20c80f6ea22a5511f1055bc762ad9] [INFO] 
[7410efc08ea20c80f6ea22a5511f1055bc762ad9] [ERROR] Errors: 
[7410efc08ea20c80f6ea22a5511f1055bc762ad9] [ERROR]   CassandraMailRepositoryTest.storingAndRemovingMessagesConcurrentlyShouldLeadToConsistentResult » Execution
```

test this please

**Comment by rouazana** [comment]
```
[7410efc08ea20c80f6ea22a5511f1055bc762ad9] [ERROR] Errors: 
[7410efc08ea20c80f6ea22a5511f1055bc762ad9] [ERROR]   JamesAppSpringMainTest.testServer:27 » BeanCreation Error creating bean with n...
[7410efc08ea20c80f6ea22a5511f1055bc762ad9] [ERROR] org.apache.james.app.spring.JamesSpringContextTest.springShouldLoadAndAddOnlyOneQuotaUpdaterListener(org.apache.james.app.spring.JamesSpringContextTest)
[7410efc08ea20c80f6ea22a5511f1055bc762ad9] [ERROR]   Run 1: JamesSpringContextTest.setup:45 » BeanCreation Error creating bean with name '...
[7410efc08ea20c80f6ea22a5511f1055bc762ad9] [ERROR]   Run 2: JamesSpringContextTest.tearDown:52 NullPointer
```

**Comment by remk** [comment]
```
[7410efc08ea20c80f6ea22a5511f1055bc762ad9] Caused by: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'taskManager' defined in class path resource [META-INF/org/apache/james/spring-server.xml]: Unsatisfied dependency expressed through constructor argument with index 0 of type [org.apache.james.task.eventsourcing.Hostname]: : No qualifying bean of type [org.apache.james.task.eventsourcing.Hostname] found for dependency: expected at least 1 bean which qualifies as autowire candidate for this dependency. Dependency annotations: {}; nested exception is org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type [org.apache.james.task.eventsourcing.Hostname] found for dependency: expected at least 1 bean which qualifies as autowire candidate for this dependency. Dependency annotations: {}
[7410efc08ea20c80f6ea22a5511f1055bc762ad9] 	at org.apache.james.app.spring.JamesAppSpringMainTest.testServer(JamesAppSpringMainTest.java:27)
[7410efc08ea20c80f6ea22a5511f1055bc762ad9] Caused by: org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type [org.apache.james.task.eventsourcing.Hostname] found for dependency: expected at least 1 bean which qualifies as autowire candidate for this dependency. Dependency annotations: {}
[7410efc08ea20c80f6ea22a5511f1055bc762ad9] 	at org.apache.james.app.spring.JamesAppSpringMainTest.testServer(JamesAppSpringMainTest.java:27)
[7410efc08ea20c80f6ea22a5511f1055bc762ad9] 
[7410efc08ea20c80f6ea22a5511f1055bc762ad9] [INFO] 
[7410efc08ea20c80f6ea22a5511f1055bc762ad9] [INFO] Results:
[7410efc08ea20c80f6ea22a5511f1055bc762ad9] [INFO] 
[7410efc08ea20c80f6ea22a5511f1055bc762ad9] [ERROR] Errors: 
[7410efc08ea20c80f6ea22a5511f1055bc762ad9] [ERROR]   JamesAppSpringMainTest.testServer:27 » BeanCreation Error creating bean with n...
[7410efc08ea20c80f6ea22a5511f1055bc762ad9] [ERROR] org.apache.james.app.spring.JamesSpringContextTest.springShouldLoadAndAddOnlyOneQuotaUpdaterListener(org.apache.james.app.spring.JamesSpringContextTest)
[7410efc08ea20c80f6ea22a5511f1055bc762ad9] [ERROR]   Run 1: JamesSpringContextTest.setup:45 » BeanCreation Error creating bean with name '...
[7410efc08ea20c80f6ea22a5511f1055bc762ad9] [ERROR]   Run 2: JamesSpringContextTest.tearDown:52 NullPointer
```

**Comment by remk** [comment]
```
[254cb4d7e75aafb5fc267172b0347b7ee689b4ce] [91mW: Failed to fetch http://archive.ubuntu.com/ubuntu/dists/bionic/InRelease  Could not connect to archive.ubuntu.com:80 (91.189.91.23), connection timed out
[254cb4d7e75aafb5fc267172b0347b7ee689b4ce] W: Failed to fetch http://archive.ubuntu.com/ubuntu/dists/bionic-updates/InRelease  Unable to connect to archive.ubuntu.com:http:
[254cb4d7e75aafb5fc267172b0347b7ee689b4ce] W: Failed to fetch http://archive.ubuntu.com/ubuntu/dists/bionic-backports/InRelease  Unable to connect to archive.ubuntu.com:http:
[254cb4d7e75aafb5fc267172b0347b7ee689b4ce] W: Some index files failed to download. They have been ignored, or old ones used instead.
[254cb4d7e75aafb5fc267172b0347b7ee689b4ce] [0mRemoving intermediate container 39ec141e0dcf
[254cb4d7e75aafb5fc267172b0347b7ee689b4ce]  ---> 1d5620f8ddff
[254cb4d7e75aafb5fc267172b0347b7ee689b4ce] Step 4/13 : RUN apt-get install -y git wget unzip
[254cb4d7e75aafb5fc267172b0347b7ee689b4ce]  ---> Running in f395ea167fc2
[254cb4d7e75aafb5fc267172b0347b7ee689b4ce] Reading package lists...
[254cb4d7e75aafb5fc267172b0347b7ee689b4ce] Building dependency tree...
[254cb4d7e75aafb5fc267172b0347b7ee689b4ce] Reading state information...
[254cb4d7e75aafb5fc267172b0347b7ee689b4ce] Package unzip is not available, but is referred to by another package.
[254cb4d7e75aafb5fc267172b0347b7ee689b4ce] This may mean that the package is missing, has been obsoleted, or
[254cb4d7e75aafb5fc267172b0347b7ee689b4ce] is only available from another source
[254cb4d7e75aafb5fc267172b0347b7ee689b4ce] 
[254cb4d7e75aafb5fc267172b0347b7ee689b4ce] [91mE: Package 'unzip' has no installation candidate
[254cb4d7e75aafb5fc267172b0347b7ee689b4ce] [0mThe command '/bin/sh -c apt-get install -y git wget unzip' returned a non-zero code: 100
[254cb4d7e75aafb5fc267172b0347b7ee689b4ce] Running: Shell Script
```
test this please

**Comment by remk** [comment]
```
[254cb4d7e75aafb5fc267172b0347b7ee689b4ce] Step 4/13 : RUN apt-get install -y git wget unzip
[254cb4d7e75aafb5fc267172b0347b7ee689b4ce]  ---> Running in b056c6cf6991
[254cb4d7e75aafb5fc267172b0347b7ee689b4ce] Reading package lists...
[254cb4d7e75aafb5fc267172b0347b7ee689b4ce] Building dependency tree...
[254cb4d7e75aafb5fc267172b0347b7ee689b4ce] Reading state information...
[254cb4d7e75aafb5fc267172b0347b7ee689b4ce] Package unzip is not available, but is referred to by another package.
[254cb4d7e75aafb5fc267172b0347b7ee689b4ce] This may mean that the package is missing, has been obsoleted, or
[254cb4d7e75aafb5fc267172b0347b7ee689b4ce] is only available from another source
[254cb4d7e75aafb5fc267172b0347b7ee689b4ce] 
[254cb4d7e75aafb5fc267172b0347b7ee689b4ce] [91mE: Package 'unzip' has no installation candidate
[254cb4d7e75aafb5fc267172b0347b7ee689b4ce] [0mThe command '/bin/sh -c apt-get install -y git wget unzip' returned a non-zero code: 100
[254cb4d7e75aafb5fc267172b0347b7ee689b4ce] Running: Shell Script
[254cb4d7e75aafb5fc267172b0347b7ee689b4ce] [566] Running shell script
```

**Comment by remk** [comment]
test this please

File: server/task-distributed/src/main/scala/org/apache/james/task/eventsourcing/distributed/TaskEventDTO.scala [hunk 7]
```diff
@@ -47,14 +47,15 @@ object CreatedDTO {
 
 case class StartedDTO(@JsonProperty("type") typeName: String,
                       @JsonProperty("aggregate") aggregateId: String,
-                      @JsonProperty("event") eventId: Int)
+                      @JsonProperty("event") eventId: Int,
+                      @JsonProperty("hostname") getHostname: String)
```
Comment (mbaechler) [inline server/task-distributed/src/main/scala/org/apache/james/task/eventsourcing/distributed/TaskEventDTO.scala]: why getHostname?

Comment (blackheaven) [reply server/task-distributed/src/main/scala/org/apache/james/task/eventsourcing/distributed/TaskEventDTO.scala]: to make it (de)serializable

**Comment by blackheaven** [comment]
```
[ERROR]   CassandraMailRepositoryTest.storingAndRemovingMessagesConcurrentlyShouldLeadToConsistentResult » Execution
```

test this please

File: server/task-distributed/src/main/scala/org/apache/james/task/eventsourcing/distributed/TaskEventDTO.scala (hunk 7 above)
Comment (blackheaven) [reply server/task-distributed/src/main/scala/org/apache/james/task/eventsourcing/distributed/TaskEventDTO.scala]: that's required by jackson to make it works

Comment (mbaechler) [reply server/task-distributed/src/main/scala/org/apache/james/task/eventsourcing/distributed/TaskEventDTO.scala]: ok

**Comment by mbaechler** [comment]
I can't approve my own PR. It's :+1: to me

**Comment by blackheaven** [comment]
```
[ERROR]   MemorySpamAssassinContractTest.deletingSpamMessageShouldNotImpactSpamAssassinState » ConditionTimeout
```

test this please

**Comment by remk** [comment]
```
[6663134404d35d535405f34d1a48b9bb2947ab0b] [INFO] Tests run: 24, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.187 s - in org.apache.james.mailetcontainer.impl.JamesMailetContextTest
[6663134404d35d535405f34d1a48b9bb2947ab0b] [INFO] Running org.apache.james.mailetcontainer.impl.matchers.OrTest
[6663134404d35d535405f34d1a48b9bb2947ab0b] [INFO] Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.021 s - in org.apache.james.mailetcontainer.impl.matchers.OrTest
[6663134404d35d535405f34d1a48b9bb2947ab0b] [INFO] Running org.apache.james.mailetcontainer.impl.matchers.XorTest
[6663134404d35d535405f34d1a48b9bb2947ab0b] [INFO] Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.008 s - in org.apache.james.mailetcontainer.impl.matchers.XorTest
[6663134404d35d535405f34d1a48b9bb2947ab0b] [INFO] Running org.apache.james.mailetcontainer.impl.matchers.AndTest
[6663134404d35d535405f34d1a48b9bb2947ab0b] [INFO] Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.002 s - in org.apache.james.mailetcontainer.impl.matchers.AndTest
[6663134404d35d535405f34d1a48b9bb2947ab0b] [INFO] Running org.apache.james.mailetcontainer.impl.matchers.NotTest
[6663134404d35d535405f34d1a48b9bb2947ab0b] [INFO] Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.003 s - in org.apache.james.mailetcontainer.impl.matchers.NotTest
[6663134404d35d535405f34d1a48b9bb2947ab0b] [INFO] 
[6663134404d35d535405f34d1a48b9bb2947ab0b] [INFO] Results:
[6663134404d35d535405f34d1a48b9bb2947ab0b] [INFO] 
[6663134404d35d535405f34d1a48b9bb2947ab0b] [ERROR] Errors: 
[6663134404d35d535405f34d1a48b9bb2947ab0b] [ERROR]   JamesMailSpoolerTest.threadSuicideShouldAckTheItem:126 » ConditionTimeout Cond...
[6663134404d35d535405f34d1a48b9bb2947ab0b] [INFO] 
```
test this please

**Comment by mbaechler** [comment]
merged

**Pull request closed without merging**